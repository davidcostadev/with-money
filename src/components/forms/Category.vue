<template>
  <form>
    <md-card>
      <md-card-content>
        <md-field>
          <label for="name">Name</label>
          <md-input v-model="nameMutable" id="name" />
        </md-field>
        <md-field>
          <label for="type">Type</label>
          <md-select v-model="typeMutable" name="type" id="type">
            <md-option v-for="option in typeOptions"
              :key="option.id"
              :value="option.id"
              :id="`type-${option.id}`"
            >
              {{option.name}}
            </md-option>
          </md-select>
        </md-field>
      </md-card-content>
      <md-card-actions>
        <md-button
        class="md-primary"
        @click="save"
        :disabled="isLoading"
        id="save"
      >Save</md-button>
      </md-card-actions>
    </md-card>
  </form>
</template>

<script>
export default {
  props: {
    name: {
      type: String,
      default: '',
    },
    type: {
      type: String,
      default: 'out',
    },
    onSave: {
      type: Function,
      required: true,
    },
  },
  data() {
    return {
      isLoading: false,
      nameMutable: this.name,
      typeMutable: this.type,
      typeOptions: [
        {
          id: 'out',
          name: 'Saida',
        },
        {
          id: 'in',
          name: 'Entrada',
        },
      ],
    };
  },
  methods: {
    async save() {
      this.isLoading = true;

      await this.onSave({
        name: this.nameMutable,
        type: this.typeMutable,
      });

      this.isLoading = false;
    },
  },
};
</script>
