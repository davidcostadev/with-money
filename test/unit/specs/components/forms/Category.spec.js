import { createLocalVue, shallowMount, mount } from '@vue/test-utils';
import VueMaterial from 'vue-material';
import Category from '../../../../../src/components/forms/Category';

const localVue = createLocalVue();

localVue.use(VueMaterial);

describe('Category Component', () => {
  describe('Render', () => {
    it('default', () => {
      const onSave = jest.fn();
      const wrapper = shallowMount(Category, {
        localVue,
        propsData: { onSave },
      });

      expect(wrapper.isVueInstance).toBeTruthy();
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Behavior', () => {
    it('onSave', () => {
      const onSave = jest.fn();
      const wrapper = mount(Category, {
        localVue,
        propsData: {
          onSave,
        },
        sync: false,
      });

      wrapper.find('#save').trigger('click');

      expect(onSave.mock.calls[0][0]).toHaveProperty('name');
      expect(onSave.mock.calls[0][0]).toHaveProperty('type');

      expect(onSave.mock.calls[0][0].name).toBe('');
      expect(onSave.mock.calls[0][0].type).toBe('out');
    });
  });
});
