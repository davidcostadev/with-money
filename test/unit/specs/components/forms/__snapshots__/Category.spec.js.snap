// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category Component Render default 1`] = `
<form>
  <mdcard-stub>
    <mdcardcontent-stub>
      <mdfield-stub>
        <label
          for="name"
        >
          Name
        </label>
         
        <mdinput-stub />
      </mdfield-stub>
       
      <mdfield-stub>
        <label
          for="type"
        >
          Type
        </label>
         
        <mdselect-stub>
          <mdoption-stub
            id="type-out"
          >
            
            Saida
          
          </mdoption-stub>
          <mdoption-stub
            id="type-in"
          >
            
            Entrada
          
          </mdoption-stub>
        </mdselect-stub>
      </mdfield-stub>
    </mdcardcontent-stub>
     
    <mdcardactions-stub>
      <mdbutton-stub
        class="md-primary"
        id="save"
      >
        Save
      </mdbutton-stub>
    </mdcardactions-stub>
  </mdcard-stub>
</form>
`;
